### DISCUSSION TITLE
useLocalStorage default value

### DISCUSSION THREAD

**Discussion by tianhuil** [mantinedev/mantine#1203] [Feature requests]
Naively using `useLocalStorage` with `superjson` generates a cryptic error:

```
SyntaxError: Unexpected token u in JSON at position 0
```

Here is the code.
```ts
useLocalStorage({
    key,
    defaultValue,
    serialize: superjson.stringify,
    deserialize: superjson.parse,
  })
```

It turns out that this is because we are trying to deserialize an undefined value from localStorage for a new key.  The problem is fairly non-obvious, especially given the unhelpful error message.  Once you have the problem, a workaround is simple:

```ts
useLocalStorage({
    key,
    defaultValue,
    serialize: superjson.stringify,
    // initial value will be undefined and this elides that issue
    deserialize: (str) =>
      str === undefined ? defaultValue : superjson.parse(str),
  })
```

I propose we either:
1. Update the docs [here](https://mantine.dev/hooks/use-local-storage/) with `superjson` as an example to showcase the trickiness of deserialization.
2. Handle the `undefined` for the user: `undefined` is almost certainly means localStorage was not initialized and `defaultValue` is the intended value.
3. Somehow catch this error and generate a warning (suboptimal).  Can be combined with (1) above.

**Comment by rtivital** [comment]
Superjson seems nice, never heard about it before. I think we can include it in documentation, do you want to create a PR with these docs?

**Comment by tianhuil** [comment]
Thanks @rtivital: here it is: https://github.com/mantinedev/mantine/compare/master...tianhuil:patch-1

**Reply by rtivital** [reply]
So, can you create a PR?

**Reply by tianhuil** [reply]
Sorry about that: https://github.com/mantinedev/mantine/pull/1207

**Reply by tianhuil** [reply]
And the above has been merged!
